### PR TITLE
helm: delete hook-related resource on re-apply

### PR DESCRIPTION
**Pull request by hugoShaka**
Follow-up on a customer support request.

In case of a previously failed deployment, this will allow users to attempt directly another chart release without having to delete the jobs. This behaviour is less awkward but might hide previous failures. However, we can legitimately expect users to monitor their failures.

### CODE REVIEW

**Comment by hugoShaka** [comment]
Note to self: I need to test the new behaviour before merging.

**Review by webvictim — APPROVED** [review]
LGTM assuming the test is OK

**Review by r0mant — APPROVED** [review]

**Review by tigrato — APPROVED** [review]

**Comment by hugoShaka** [comment]
Test is successful:

I created a broken release with an invalid version (job's pod is stuck in imagePullBackOff)

```shell
shaka@Hugos-MacBook-Pro ~/g/s/g/g/teleport (hugo/chart-do-not-block-on-failed-hook)> helm ls -n teleport-cluster
NAME            NAMESPACE               REVISION        UPDATED                                 STATUS  CHART                           APP VERSION
teleport        teleport-cluster        5               2023-04-04 13:36:01.874112 -0400 EDT    failed  teleport-cluster-13.0.0-dev     13.0.0-dev 
shaka@Hugos-MacBook-Pro ~/g/s/g/g/teleport (hugo/chart-do-not-block-on-failed-hook)> kubectl get cm -n teleport-cluster
NAME                  DATA   AGE
kube-root-ca.crt      1      13d
teleport-auth         2      13d
teleport-auth-test    2      42s
teleport-proxy        1      13d
teleport-proxy-test   1      42s
shaka@Hugos-MacBook-Pro ~/g/s/g/g/teleport (hugo/chart-do-not-block-on-failed-hook)> kubectl get jobs -n teleport-cluster
NAME                 COMPLETIONS   DURATION   AGE
teleport-auth-test   0/1           49s        49s
```

Then successfully re-applied with a valid version without deleting failed artifacts:

```shell
shaka@Hugos-MacBook-Pro ~/g/s/g/g/teleport (hugo/chart-do-not-block-on-failed-hook)> helm upgrade teleport . -n teleport-cluster -f ~/playground/cloud-test/gke/gcp-values.yaml --set teleportVersionOverride=12.1.2
```

There are no traces left of a broken release nor tests jobs:
```shell
shaka@Hugos-MacBook-Pro ~/g/s/g/g/teleport (hugo/chart-do-not-block-on-failed-hook)> kubectl get cm -n teleport-cluster
NAME               DATA   AGE
kube-root-ca.crt   1      13d
teleport-auth      2      13d
teleport-proxy     1      13d
shaka@Hugos-MacBook-Pro ~/g/s/g/g/teleport (hugo/chart-do-not-block-on-failed-hook)> kubectl get jobs -n teleport-cluster
No resources found in teleport-cluster namespace.
```

**Comment by webvictim** [comment]
This is super useful; I just ran into this exact issue today with a prospect.

**Comment by emoshaya** [comment]
We are having the following issue https://github.com/gravitational/teleport/issues/24035

The hook.yaml and delete_hook.yaml are both missing the nodeSelector block and are trying to schedule pods on a Windows node in our cluster even though we've got nodeSelector set to linux.

**Comment by webvictim** [comment]
I'll fix the `nodeSelector` issue in #24012

**Pull request merged**
